Zend/zend_vm_def.h

```diff
@@ -8402,6 +8402,16 @@ ZEND_VM_COLD_CONSTCONST_HANDLER(188, ZEND_SWITCH_STRING, CONST|TMPVARCV, CONST,
 	}
 }
 
+ZEND_VM_COLD_CONSTCONST_HANDLER(195, ZEND_MATCH_LONG, CONST|TMPVARCV, CONST, JMP_ADDR)
+{
+	ZEND_VM_DISPATCH_TO_HANDLER(ZEND_SWITCH_LONG);
+}
+
+ZEND_VM_COLD_CONSTCONST_HANDLER(196, ZEND_MATCH_STRING, CONST|TMPVARCV, CONST, JMP_ADDR)
+{
+	ZEND_VM_DISPATCH_TO_HANDLER(ZEND_SWITCH_STRING);
```

Comment:
I'd have assumed something like the following would work, and be more efficient in the `default` case as discussed earlier
(and analogous changes to ZEND_SWITCH_LONG for ZEND_MATCH_LONG

```c
	USE_OPLINE
	zval *op, *jump_zv;
	HashTable *jumptable;

	op = GET_OP1_ZVAL_PTR_UNDEF(BP_VAR_R);

	if (Z_TYPE_P(op) != IS_STRING) {
		if (OP1_TYPE == IS_CONST) {
			/* default */
			ZEND_VM_C_GOTO(default_branch);
		} else {
			ZVAL_DEREF(op);
			if (Z_TYPE_P(op) != IS_STRING) {
				/* default */
				ZEND_VM_C_GOTO(default_branch);
			}
		}
	}

	jumptable = Z_ARRVAL_P(GET_OP2_ZVAL_PTR(BP_VAR_R));
	jump_zv = zend_hash_find_ex(jumptable, Z_STR_P(op), OP1_TYPE == IS_CONST);
	if (jump_zv != NULL) {
		ZEND_VM_SET_RELATIVE_OPCODE(opline, Z_LVAL_P(jump_zv));
		ZEND_VM_CONTINUE();
	} else {
ZEND_VM_C_LABEL(default_branch):
		/* default */
		ZEND_VM_SET_RELATIVE_OPCODE(opline, opline->extended_value);
		ZEND_VM_CONTINUE();
	}
```

Reply:
Yeah, this should be possible for the `match`. It wouldn't work for `switch` right now as the `ZEND_SWITCH_*` opcodes do strict type checking and do loose comparison through the `IS_EQUAL`/`JMPNZ` chain.

Reply:
Done. In the optimal case the opcodes now look like this:

```
0001 MATCH_LONG CV0($char) 1: 0002, 2: 0003, ..., default: 0008
0002 RETURN string("1")
0003 RETURN string("2, 3")
...
0008 RETURN string("default")
```